### DISCUSSION TITLE
拡張APIについて

### DISCUSSION THREAD

**Discussion by bdunderscore** [bdunderscore/modular-avatar#124] [Ideas]
Modular Avatar自体に入るほど汎用的でないコンポーネントにも対応すべく、拡張APIを敷設することを考えています。

MAと同様にVRCSDK自体の拡張APIを使えばいいのでは？という突っ込みはありそうだけど、
①自作コンポーネントを付けたままビルドを可能にするのが面倒
②プレイモードへの対応が面倒
③一時的アセットの管理が面倒
④Manual bakeへの介入が必要になる
⑤MAの処理途中に割り込む必要があるかもしれない（例：今後最適化を実装する場合は、最適化前に介入する必要があるなど）

拡張APIを作ることに当たって解決すべき点をとりあえずリストアップすると

①拡張アセットの処理パスの実行順管理が必要。また、デバッグように、拡張を切る必要があったりなど
②エラーが発生した時、ユーザーにとって分かりやすく表示（コンソールを読まない人がほとんど）
③メッセージ翻訳システムも拡張させるべき
④Modular Avatar内包コンポーネントと外部の拡張の境界を明記する必要がある
⑤必ずしも必須ではない拡張コンポーネントを配布アセットに入れた場合、ミッシングコンポーネントエラーにならないための施しもほしい

…まぁ、④に関してはFAQやドキュメンテーションにロゴを使わないでって明記する感じで行くとして、①と②から着手するべき感じですね。

**Comment by bdunderscore** [comment]
【処理パス管理】

スケッチとして、

```
internal class MyExtension : ModularAvatarExtension {
  public override string DisplayName => "MyExtension";
  public override string PackageName => "com.foo.my-extension";
  public override string Version => "1.0.0";


  public void RegisterPasses(PassInfo passes) {
    passes.AddPass("GenerateAnimation", MyGenerateAnimationPass.ProcessAvatar)
      // 指定したパスを前提し、その前に処理する
      .RequireBefore("ModularAvatar", "MergeAnimatorProcessor")
      // 指定したパスを前提とし、その後に処理する
      .Require("ModularAvatar", "RenameParameters")
      // 指定したパスを前提としないが、あった場合はその後に処理する
      .AfterIfPresent("OtherExtension", "SomeOtherPass");
  }
}

internal class MyGenerateAnimationPass {
  public static void ProcessAvatar(CompileJob job) {
    var avatar = job.Avatar;
    if (SomethingWentWrong) {
      job.Error(...);
    }
  }
}
```

他にも、現在一部の情報がstatic変数で管理されるが、同じように`CompileJob`に移すことで初期か忘れを防止できます。
```
// ...
  var myData = compileJob.GetState<MyType>(); // default(MyType)に初期化される
// ...
  compileJob.SetState<MyType>(myData);
```

**Reply by bdunderscore** [reply]
問題点
・パス順番の指定で、実行されるかもしれない位置が複数ある場合、どうやって選ぶか？拡張作者のテストが不十分になる恐れがある

**Comment by bdunderscore** [comment]
【エラー報告】

```
// localization file
  "myNamespace.error.ErrorCode" => "設定hogeが{0}になっているぞ！"

// エラー発生位置(component)を記録、GUIで飛べるように
// 実行中のpassから問題の拡張を指摘
throw new CompileAvatarError("ErrorCode", component.hoge, component); 

job.reportError("ErrorCode", component.hoge, component); // 実行を途中中断せずに、他のエラーも集めて最後に失敗させる
job.reportWarning(...); // 中断はしないが、注意を表示

```

**Comment by bdunderscore** [comment]
・MAEditorBaseはpublic化しない
・Util/RuntimeUtilをどこまで公開するかを考えること

**Reply by raiti-chan** [reply]
`MAEditorBase`の代わりに`InspectorCommon.DisplayOutOfAvatarWarning`の公開、もしくはそれに準ずる機能を追加できる関数が欲しいですね。

**Comment by bdunderscore** [comment]
【Missing Component対策】
`MA Optional Part`みたいな、前提コンポーネントがあった場合のみ指定したプレハブをスポーンさせるコンポーネントとか？
少し挙動が複雑なので使い勝手がどうなるかが不安
